packages/itwin/tree-widget/src/components/TreeWidgetComponent.tsx

```diff
@@ -5,31 +5,129 @@
 
 import "./TreeWidgetComponent.scss";
 import * as React from "react";
-import { SelectableContent, SelectableContentDefinition } from "@itwin/components-react";
+import { useActiveIModelConnection } from "@itwin/appui-react";
+import { SelectableContent, SelectableContentDefinition, SelectableContentProps } from "@itwin/components-react";
+import { IModelConnection } from "@itwin/core-frontend";
+import { FillCentered } from "@itwin/core-react";
+import { ProgressLinear } from "@itwin/itwinui-react";
+import { TreeWidget } from "../TreeWidget";
 
 export interface TreeDefinition {
+  /** Id of the tree */
   id: string;
+  /** Callback that is used to get tree label */
   getLabel: () => string;
+  /** Callback that is used to render tree component */
   render: () => React.ReactNode;
+  /**
+   * Callback that is used to determine if tree should be shown for current active iModel connection.
+   * If callback is `undefined` tree is shown for all iModel connections.
+   */
+  shouldShow?: (imodel: IModelConnection) => Promise<boolean>;
 }
 
 interface TreeWidgetComponentProps {
   trees: TreeDefinition[];
 }
 
 export function TreeWidgetComponent(props: TreeWidgetComponentProps) {
-  const trees: SelectableContentDefinition[] = props.trees.map((tree) => ({
-    id: tree.id,
-    label: tree.getLabel(),
-    render: tree.render,
-  }));
+  const { trees: treeDefinitions } = props;
+  const trees = useActiveTrees(treeDefinitions);
 
   return (
     <div className="tree-widget-visibility-widget">
-      <SelectableContent
-        children={trees} // eslint-disable-line react/no-children-prop
-        defaultSelectedContentId={trees[0].id}
-      />
+      <SelectableContent {...getSelectableContentProps(trees)} />
     </div>
   );
 }
+
+function useActiveTrees(treeDefinitions: TreeDefinition[]) {
+  const [trees, setTrees] = React.useState<SelectableContentDefinition[]>();
+  const imodel = useActiveIModelConnection();
+
+  React.useEffect(() => {
+    if (!imodel)
+      return;
+
+    let disposed = false;
+    (async () => {
+      const visibleTrees = await getActiveTrees(treeDefinitions, imodel);
+      if (!disposed) {
+        setTrees(visibleTrees);
+      }
+    })();
+
+    return () => { disposed = true; };
+  }, [treeDefinitions, imodel]);
+
+  return trees;
+}
+
+async function getActiveTrees(treeDefinitions: TreeDefinition[], imodel: IModelConnection): Promise<SelectableContentDefinition[]> {
+  const activeTrees: SelectableContentDefinition[] = [];
+  for (const treeDef of treeDefinitions) {
+    if (treeDef.shouldShow !== undefined && !(await treeDef.shouldShow(imodel))) {
+      continue;
+    }
```

Comment:
Based on the fact that some of these checks are going to involve sending an HTTP request, maybe it makes sense to send them all at once (`Promise.all`)?